File: Analyzer/symtable.cpp
#include "symtable.h"
#include "Visitor.h"
#include "error.h"

inline Type *ConstDeclToType(ConstDecl *decl) {
  return new Constant(decl->type);
}

Type *VarDeclToType(VarDecl *decl) {
  if (decl->is_array) {
    assert(decl->size > 0);
    return new Array(decl->type, decl->size);
  }
  else {
    return new Variable(decl->type);
  }
}

Type *FuncDefToType(FuncDef *decl) {
  auto fun = new Function(decl->return_type, {});
  for (auto arg: decl->args) {
    fun->args.push_back(arg->type);
  }
  return fun;
}

// convert a declaration to type
Type *DeclToType(Decl *decl) {
  if (auto x = subclass_cast<ConstDecl>(decl))
    return ConstDeclToType(x);
  if (auto x = subclass_cast<VarDecl>(decl))
    return VarDeclToType(x);
  return FuncDefToType(subclass_cast<FuncDef>(decl));
}

inline Type *DeclToType(Arg *arg) {
  return new Variable(arg->type);
}

// Define a declaration globally.
void DefineGlobalDecl(Decl *decl, TableType &global, ErrorManager &e) {
  if (global.count(decl->name)) {
    e.NameError(decl->loc, "redefinition of identifier", Quote(decl->name), "in <module>");
    return;
  }
  global.emplace(std::make_pair(decl->name, Entry(DeclToType(decl), Scope::Global, decl->loc, decl->name)));
}

// Define a declaration locally.
// DeclLike takes Decl or Arg
template <typename DeclLike>
void DefineLocalDecl(DeclLike *decl, TableType &local, const TableType &global,
                     ErrorManager &e, const String &funcname) {
  auto where = "in function " + Quote(funcname);
  if (local.count(decl->name)) {
    e.NameError(decl->loc,
        "redefinition of identifier", Quote(decl->name), where);
    return;
  }
  if (auto iter = global.find(decl->name); iter != global.end() && IsInstance<Function>(iter->second.type)) {
    e.NameError(decl->loc,
        "local identifier", Quote(decl->name), where, "shallows a global function name");
    return;
  }
  local.emplace(std::make_pair(decl->name, Entry(DeclToType(decl), Scope::Local, decl->loc, decl->name)));
}

// Enter all global const/var declarations into dict
void MakeGlobal(Program *prog, TableType &dict, ErrorManager &e) {
  for (auto decl: prog->decls) {
    if (IsInstance<FuncDef>(decl))
      continue; // not now
    DefineGlobalDecl(decl, dict, e);
  }
}

// Visitor that resolves local names for a function
class LocalResolver: public VisitorBase<LocalResolver>,
                     public ChildrenVisitor<LocalResolver> {
public:
  // The Ast of the function
  FuncDef *fun;
  // Global namespace to fail back
  const TableType &global;
  // Local namespace with local declarations and function arguments
  TableType &local;
  // Report errors
  ErrorManager &e;

  LocalResolver(FuncDef *fun,
      const TableType &global, TableType &local, ErrorManager &e):
    fun(fun), global(global), local(local), e(e) {}

  // Resolve one name
  void ResolveName(const String &name, const Location &loc) {
    if (local.find(name) != local.end())
      return; // already defined locally
    // define globally
    if (auto x = global.find(name); x != global.end()) {
      // make a copy of the global entry
      Entry local_entry(x->second);
      local_entry.location = loc; // use local location
      local.emplace(std::make_pair(name, std::move(local_entry)));
    }
    else {
      e.NameError(loc,
          "undefined identifier", Quote(name), "in function", Quote(fun->name));
    }
  }

  // pull in VisitorBase's methods
  void visitStmt(Stmt *s) {
    return VisitorBase::visitStmt<void>(s);
  }

  void visitExpr(Expr *s) {
    return VisitorBase::visitExpr<void>(s);
  }

  void visitCall(Call *x) {
    ResolveName(x->func, x->loc);
    ChildrenVisitor::visitCall(x);
  }

  // these do not have identifiers
  void visitNum(Num *x) {}
  void visitStr(Str *x) {}
  void visitChar(Char *x) {}

  void visitSubscript(Subscript *x) {
    ResolveName(x->name, x->loc);
    ChildrenVisitor::visitSubscript(x);
  }

  void visitName(Name *x) {
    ResolveName(x->id, x->loc);
  }

  // public interface
  void Resolve() {
    for (auto stmt: fun->stmts) {
      visitStmt(stmt);
    }
  }
};

// define and resolve names of a function
void MakeLocal(FuncDef *fun,
    TableType &top, TableType &local, ErrorManager &e) {
  // define fun itself in global first
  DefineGlobalDecl(fun, top, e);

  // define arguments of a function
  for (auto arg: fun->args) {
    DefineLocalDecl(arg, local, top, e, fun->name);
  }

  // define const/var declarations of a function
  auto where = "function " + Quote(fun->name);
  for (auto decl: fun->decls) {
    DefineLocalDecl(decl, local, top, e, fun->name);
  }

  // resolve local names
  LocalResolver resolver(fun, top, local, e);
  resolver.Resolve();
  // errors left in ErrorManager
}

// public interface
bool BuildSymbolTable(Program *prog, SymbolTable &table) {
  ErrorManager e;
  // build global table first
  auto &global = table.global;
  // no early return to check more errors
  MakeGlobal(prog, global, e);

  // visit all FuncDef and build their local tables
  auto &locals = table.locals;
  for (auto decl: prog->decls) {
    if (auto fun = subclass_cast<FuncDef>(decl)) {
      TableType local;
      MakeLocal(fun, global, local, e);
      if (e.IsOk()) {
        locals.emplace(std::make_pair(fun->name, std::move(local)));
      }
    }
  }
  return e.IsOk();
}

void CheckTable(const TableType &table) {
  for (const auto &item: table) {
    assert(item.first == item.second.name);
    assert(item.second.type);
  }
}

// consistency check
void SymbolTable::Check() const {
  CheckTable(global);
  for (const auto &kv: locals) {
    CheckTable(kv.second);
    for (const auto &kv2: kv.second) {
      assert(kv2.second.scope == Scope::Local ||
          global.find(kv2.first) != global.end() &&
          "entry in local with global scope must be present in global");
    }
  }
}

const Entry &SymbolTable::LookupLocal(const String &fun, const String &name) const {
  auto iter = locals.find(fun);
  assert(iter != locals.end() && "absent function name in locals");
  const auto &local = iter->second;
  const auto &entry_iter = local.find(name);
  assert(entry_iter != local.end() && "absent name in local");
  return entry_iter->second;
}

const Entry &SymbolTable::LookupGlobal(const String &name) const {
  auto iter = global.find(name);
  assert(iter != global.end() && "absent name in global");
  return iter->second;
}

// Overloads to print various data structures
std::ostream &operator<<(std::ostream &os, Scope s) {
  switch (s) {
    case Scope::Global:
      return os << "Scope::Global";
    case Scope::Local:
      return os << "Scope::Local";
  }
}

std::ostream &operator<<(std::ostream &os, const Entry& e) {
  os << "Entry(";
  os << "type=" << e.type->ClassName() << ", "
    << "scope=" << e.scope << ", "
    << "location=" << e.location << ", "
    << "name=" << e.name;
  return os << ")";
}

// print a dict with string key and generic value
template <typename Value>
std::ostream &operator<<(
    std::ostream &os, const std::unordered_map<String, Value> &t) {
  int i = 0;
  int size = t.size();

  os << "{";
  for (const auto &b: t) {
    os << Quote(b.first) << ": " << b.second;
    if (i != size - 1) {
      os << ", ";
    }
    i++;
  }
  return os << "}";
}

std::ostream &operator<<(std::ostream &os, const SymbolTable &t) {
  os << "SymbolTable(";
  os << "global=" << t.global << ", "
    << "locals=" << t.locals;
  return os << ")";
}

File: simplecompiler/compiler/symtable.py
"""Symbol table construction"""

from functools import singledispatch
from collections import namedtuple
from enum import Enum
from operator import itemgetter
from operator import attrgetter
from pprint import pprint

# For generating tests
from string import Template
import re

from simplecompiler.compiler.AST import *
from simplecompiler.util import ErrorManager
from simplecompiler.util import format_code
Arg = arg

__all__ = ["build_symtable", "SymbolTable", "symtable_make_test"]


# An Entry represents a single resolved symbol
Entry = namedtuple("Entry", "type scope loc name")
Scope = Enum("Scope", "Local Global")
MODULE_NAME = "<module>"


class Type:
    def __repr__(self):
        return self.__class__.__name__

# pure data class to hold type information


class Constant(Type):

    def __init__(self, type):
        self.type = type


class Variable(Type):

    def __init__(self, type):
        self.type = type


class Array(Type):

    def __init__(self, elemtype, size):
        self.elemtype = elemtype
        self.size = size


class Function(Type):

    def __init__(self, return_type, args):
        self.return_type = return_type
        self.args = args

# Convert type declaration to Types


@singledispatch
def decl_to_type(decl):
    raise TypeError("unknown decl {}".format(decl.__class__.__name__))


@decl_to_type.register(ConstDecl)
def ConstDecl2Type(decl):
    return Constant(decl.type)


@decl_to_type.register(VarDecl)
def VarDecl2Type(decl):
    if decl.is_array:
        return Array(decl.type, decl.size)
    else:
        return Variable(decl.type)


@decl_to_type.register(FuncDef)
def FuncDef2Type(decl):
    # discard names
    args = list(map(itemgetter(0), decl.args))
    return Function(decl.return_type, args)


def define_decl(decl, scope, dict, e, where):
    if decl.name in dict:
        msg = "redefinition of identifier {!r} in {}".format(decl.name, where)
        e.error(msg, decl.loc)
    else:
        type = decl_to_type(decl)
        dict[decl.name] = Entry(type, scope, decl.loc, decl.name)


def define_arg(arg, dict, e, funcname):
    if arg.name in dict:
        msg = "redefinition of argument {!r}".format(arg.name)
        e.error(msg, arg.loc)
    else:
        type = Variable(arg.type)
        dict[arg.name] = Entry(type, Scope.Local, arg.loc, arg.name)


def make_global(prog, e):
    top = {}
    for decl in prog.decls:
        if isinstance(decl, FuncDef):
            continue
        define_decl(decl, Scope.Global, top, e, MODULE_NAME)
    return top


class LocalResolver:

    def __init__(self, fun, top, local, e):
        self.fun = fun
        self.top = top
        self.local = local
        self.e = e

    def resolve_name(self, name, loc):
        if name in self.local:
            return
        if name in self.top:
            type, scope, _, name = self.top[name]
            self.local[name] = Entry(type, scope, loc, name)
        else:
            msg = "undefined identifier {!r} in function {!r}".format(
                    name, self.fun.name)
            self.e.error(msg, loc)

    def resolve(self):
        for name, loc in iter_names(self.fun):
            self.resolve_name(name, loc)



def make_local(fun, top, e):
    local = {}
    define_decl(fun, Scope.Global, top, e, MODULE_NAME)

    # insert all arguments first
    for arg in fun.args:
        define_arg(arg, local, e, fun.name)

    # then local const/var decl
    where = "function {!r}".format(fun.name)
    for decl in fun.decls:
        define_decl(decl, Scope.Local, local, e, where)

    LocalResolver(fun, top, local, e).resolve()
    return local


def iter_names(node):
    # XXX: this must cover all Ast node that has names
    # handle terminal nodes that has a name:
    if isinstance(node, (ConstDecl, VarDecl, Arg)):
        yield node.name, node.loc
    elif isinstance(node, (Read,)):
        for name in node.names:
            yield name.id, node.loc
    elif isinstance(node, Call):
        yield node.func, node.loc
        for arg in node.args:
            yield from iter_names(arg)
    elif isinstance(node, Name):
        yield node.id, node.loc
    elif isinstance(node, Subscript):
        yield node.name, node.loc
    else:
        # handle non-terminal nodes
        for child in node:
            if isinstance(child, AST):
                yield from iter_names(child)


class SymbolTable:
    """All-in-one container of all names and their information within a program.
    """

    def __init__(self, global_, locals_):
        self.global_ = global_
        self.locals_ = locals_

    def report(self, output):
        print("Global:", file=output)
        global_ = sorted(self.global_.values(), key=attrgetter('loc'))
        pprint(global_, stream=output)
        print(file=output)  # make a blank line
        print("Locals:", file=output)
        pprint(self.locals_, stream=output)


def build_symtable(prog):
    """Build a SymbolTable for the whole program"""
    assert isinstance(prog, Program)
    e = ErrorManager()
    top = make_global(prog, e)
    locals_ = {}

    for decl in prog.decls:
        if isinstance(decl, FuncDef):
            locals_[decl.name] = make_local(decl, top, e)

    return SymbolTable(top, locals_) if e.is_ok else None


def pyenum2cppenum(enum):
    # replace dot with ::
    return re.sub(r"\.", "::", str(enum))


class CppTestTempalte:
    impl = Template("""
#include "symtable.h"
#include <cassert>

void TestSymbolTable(SymbolTable& symtable) {
    auto &global = symtable.global;
    auto &locals = symtable.locals;
$test_global
$test_locals
}
""")

    testEntry = Template("""
{
assert($table.count("$name"));
const SymbolEntry &e = $table.at("$name");
assert(e.Is$class_name());
assert(e.GetScope() == $scope);
assert(e.GetName() == "$name");
}
""")

    testGlobal = Template("""
{
assert(global.size() == $size);
$test_entries
}
""")

    testLocal = Template("""
assert(locals.find("$name") != locals.end());
{
auto &local = locals["$name"];
assert(local.size() == $size);
$test_entries
}
""")

    def make_testEntry(self, e, table):
        assert isinstance(e, Entry)
        return self.testEntry.substitute(
            name=e.name,
            class_name=e.type.__class__.__name__,
            scope=pyenum2cppenum(e.scope),
            loc=", ".join(map(str, e.loc)),
            table=table,
        )

    def make_testLocal(self, name, local):
        return self.testLocal.substitute(
            name=name,
            size=len(local),
            test_entries="".join(self.make_testEntry(e, "local")
                                 for e in local.values()),
        )

    def make_testGlobal(self, global_):
        return self.testGlobal.substitute(
            size=len(global_),
            test_entries="".join(self.make_testEntry(e, "global")
                                 for e in global_.values()),
        )

    def substitute(self, symtable):
        return self.impl.substitute(
            test_global=self.make_testGlobal(symtable.global_),
            test_locals="".join(self.make_testLocal(name, local)
                                for name, local in symtable.locals_.items()),
        )


def symtable_make_test(symtable, file):
    """Generate a cpp test file using symtable"""
    code = format_code(CppTestTempalte().substitute(symtable))
    print(code, file=file)
